### ISSUE TITLE
[Feature Request] Gaussian Mixture Distribution

### ISSUE THREAD

**Issue by 09tangriro** [DLR-RM/stable-baselines3#451]
### 🚀 Feature

Add a (diagonal) GMM distribution class to `distributions.py`.

### Motivation

For now, continuous action spaces are handled by the normal distribution. However, there is no guarantee (I think) that the optimal policy conforms to a Gaussian distribution (e.g. for stochastic environments where two or more options might be equally valid), and so forcing to fit to a Gaussian could lead to high bias. A GMM would overcome this through its inherent flexibility, although likely at the cost of longer convergence time.

### Pitch

I want to add a `GMM(Distribution)` class that will implement a Gaussian Mixture distribution for continuous spaces. Furthermore, I want to add extended classes including the squashed output and gSDE implementations. I will model each Gaussian as diagonal for mathematical/computational simplicity. 
 
### Alternatives

N/A

### Additional context

N/A

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Interesting suggestion. I agree mono-gaussians are very restrictive. However is there any work showing/testing GMMs in this case? I am not quite sold by the "action distribution might not be uni-modal". I think there is no formal proof of this, but I _guess_ most situations have one or two "optimal" actions and anything around them are "ok" actions. I am also not certain about the headaches GMMs could introduce, e.g. component weights do not always zero-out when things are uni-modal (seen this happen in e.g. content generation stuff), and then there are issues with near-singular covariances, granted we can fix the variances.

TL;DR: It would be reassuring to see preliminary results on this. If this has not been studied before, it could be a perfect fit for a workshop paper :)

**Comment by 09tangriro** [comment]
Yeah I don't think this has been studied properly before (as far as I can tell anyway). This might be a silly question, but what is a workshop paper and how would I go about doing one? I haven't done a PhD (yet!) but I'd love to get into one some day and I tend to do a lot of my own research on the side (most recently tried combining Evolutionary Strategy with the Adam update equation which was interesting) but I'm unsure how to get any of this research out there 😅.

Another potentially more on topic question: for the `DiagGaussianDistribution`, why do we use a state-independent parameter for the log std of the policy instead of hooking up the log std as an output layer of the network itself? I would have thought the latter option would lead to learning more complex relationships between states and sampling noise with better performance (e.g. maybe the agent is less certain on the behaviour of one state over another because it hasn't visited it as much yet, so it assigns a higher std to the less certain state).

**Comment by Miffyli** [comment]
>  This might be a silly question, but what is a workshop paper and how would I go about doing one?

No problem :). Workshop papers are small works of research, and something like this could be a fit for a suitable workshop (they have specific topics). I recommend you find yourself a supervisor who guides and supports you during this work, or at least some other people who are willing to help. [This Discord](https://discord.com/invite/xhfNqQv) has bunch of people who will be willing to answer questions.

> Another potentially more on topic question: for the DiagGaussianDistribution, why do we use a state-independent parameter for the log std of the policy instead of hooking up the log std as an output layer of the network itself? 

As far as I know, this is precisely for the reason to avoid std becoming too small (the training tends to do that), and emperically you get better results this way.

**Comment by 09tangriro** [comment]
Ah I see, thanks for the answers to both questions they're super helpful 😊

**Comment by araffin** [comment]
Hello,

>However, there is no guarantee (I think) that the optimal policy conforms to a Gaussian distribution (e.g. for stochastic environments where two or more options might be equally valid

that's true that there is no garantee that your optimal policy will be unimodal. However, in practice, the Gaussian distribution usually gives good results.
One intuition about that is that it is probably better to learn to do one thing good (unimodal) rather than learn to do many things ok (multi modal).

There have been some work around that idea though.
Originally, [Soft Q-Learning](https://arxiv.org/abs/1702.08165) (SQL, the predecessor of SAC) was all about learning all those modalities.
In fact, the first version of [SAC](https://arxiv.org/abs/1801.01290) uses a Gaussian Mixture distribution, but it was later on replaced by a simpler diagonal Gaussian distribution.

Link to the first version of the paper: https://arxiv.org/pdf/1801.01290v1.pdf


> hy do we use a state-independent parameter for the log std of the policy instead of hooking up the log std as an output layer of the network itself?

for the std depending on observation or not: https://github.com/hill-a/stable-baselines/issues/652
